File: Wayne/WayneCompiler/src/Dfir_representation/ILeftShiftRegister.java
package Dfir_representation;

import java.util.Vector;


public class ILeftShiftRegister extends IBorderNode{
	   
	   private AssociatedRightShiftRegister associatedRightShiftRegister;
	   private Vector<Boolean> isuseds;
	   private Vector<Boolean> isDeclareds;
	   public ILeftShiftRegister(int nodeId, int parentId, Vector<ITerminal> inputTerminals,Vector<ITerminal> outputTerminals, AssociatedRightShiftRegister associatedRightShiftRegister){
		   super(nodeId,parentId,inputTerminals,outputTerminals);
		   this.associatedRightShiftRegister = associatedRightShiftRegister;
		   JNodeType = "ILeftShiftRegister";
		   isuseds = new Vector<Boolean>();
		   isDeclareds = new Vector<Boolean>();
		   for(int i=0; i<this.inputTerminals.size(); i++){
			   isuseds.addElement(false);
			   isDeclareds.addElement(false);
		   }
	   }
	   public Boolean getIsDeclared(int index){
			return this.isDeclareds.get(index);
		}
	   public void setIsDeclared(Boolean b,int index){
			this.isDeclareds.setElementAt(b, index);
		}
	   public AssociatedRightShiftRegister getAssociatedRightShiftRegister(){
		   return this.associatedRightShiftRegister;
	   }
	   public void decIndegree(){
		   this.indegree--;
		   if(this.indegree == 0) JNode.IdToJNode.get(this.parentId).indegree--;
	   }
	   public JNode getParentJNode(){
		   return JNode.IdToJNode.get(this.parentId);
	   }
	   public void setIsused(Boolean isused, int index){
		   this.isuseds.set(index, isused);
	   }
	   public String getName(int index){
		   String res;
		   res = this.getJNodeType() + Integer.toString(this.getNodeId()) + "_" + Integer.toString(index);
		   return res;
	   }
	   public Vector<Boolean> getIsuseds(){
		   return this.isuseds;
	   }
	   
	   public void print(){
		   System.out.println("ILeftShiftRegister begin");
		   System.out.println("NodeId: " + this.nodeId + ' ' + "ParentId: " + this.parentId);
		   associatedRightShiftRegister.print();
		   for(int i=0; i<inputTerminals.size(); i++){
				   inputTerminals.get(i).print();
		   }
		  for(int i=0; i<outputTerminals.size(); i++){
				   outputTerminals.get(i).print();
		   }   
		  System.out.println("ILeftShiftRegister end\n");
	   }
}

File: Jasmine/JasmineCompiler/Src/DFIR_Compiler/DFIR_Compiler/ILeftShiftRegister.cs
using System;
using System.Collections.Generic;


namespace DFIR_Compiler
{
    public class ILeftShiftRegister:IBorderNode
    {
        private AssociatedRightShiftRegister associatedRightShiftRegister;
        private List<bool> isOccupieds;
        private List<bool> isDeclareds;
        public ILeftShiftRegister(int nodeId, int parentId, List<ITerminal> inputTerminals, List<ITerminal> outputTerminals, AssociatedRightShiftRegister _associatedRightShiftRegister) : base(nodeId, parentId, inputTerminals, outputTerminals)
        {
            associatedRightShiftRegister = _associatedRightShiftRegister;            
            isOccupieds = new List<bool>();
            isDeclareds = new List<bool>();
            for (int i = 0; i < this.inputTerminals.Count; i++)
            {
                isOccupieds.Add(false);
                isDeclareds.Add(false);
            }
            INodeType = "ILeftShiftRegister";
        }


        public virtual AssociatedRightShiftRegister GetAssociatedRightShiftRegister()
        {
            return associatedRightShiftRegister;
        }
        public virtual List<bool> GetIsOccupieds()
        {
            return isOccupieds;
        }
        public virtual bool GetIsDeclared(int index)
        {
            return isDeclareds[index];
        }
        public override INode GetParentINode()
        {
            return IdToINode[parentId];
        }
        public virtual string GetName(int index)
        {
            string lsname;
            lsname = GetINodeType() + Convert.ToString(GetNodeId()) + "_" + Convert.ToString(index);
            return lsname;
        }


        public virtual void SetIsOccupied(bool isoccupied, int index)
        {
            isOccupieds[index] = isoccupied;
        }
        public virtual void SetIsDeclared(bool isdeclared, int index)
        {
            isDeclareds[index] = isdeclared;
        }


        public override void DecreaseIndegree()
        {
            indegree--;
            if (indegree == 0)
            {
                IdToINode[parentId].indegree--;
            }
        }


        public override void Print()
        {
            Console.WriteLine("------------------------------------");
            Console.WriteLine("ILeftShiftRegister:");
            Console.WriteLine("NodeId: " + nodeId + ' ' + "ParentId: " + parentId);
            associatedRightShiftRegister.Print();
            for (int i = 0; i < inputTerminals.Count; i++)
            {
                inputTerminals[i].Print();
            }
            for (int i = 0; i < outputTerminals.Count; i++)
            {
                outputTerminals[i].Print();
            }
        }

    }
}
